db/change_cache.go

```diff
@@ -732,13 +740,15 @@ func (c *changeCache) processPrincipalDoc(ctx context.Context, docID string, doc
 	base.InfofCtx(ctx, base.KeyChanges, "Received #%d (%q)", change.Sequence, base.UD(change.DocID))
 
 	changedChannels := c.processEntry(ctx, change)
-	if c.notifyChange != nil && len(changedChannels) > 0 {
-		c.notifyChange(ctx, changedChannels)
+	channelSet := channels.SetFromArrayNoValidate(changedChannels)
+
+	if c.notifyChange != nil && len(channelSet) > 0 {
+		c.notifyChange(ctx, channelSet)
 	}
 }
 
 // Handles a newly-arrived LogEntry.
```

Comment:
```suggestion
// processEntry handles newly-arrived LogEntry and returns the channels affected by this revision. This will be any existing, removed, or added channels. Channels can exist multiple times in this array.
```

I'm not sure that's the best description.

But that left me with another idea, what if you do the deduplication inside `processEntry` but no inside the `lock`? You run the function as is that produces duplicates and then deduplicate before returning to the user? I don't know if this makes any code easier to read.

Reply:
I agree with the comment needing updating, I've changed this. 

With the deduplicate in `processEntry` but outside the mutex I feel it makes sense when calling the function not in it. I know if we did inside it would deduplicate some code but I think I'd personally prefer to keep the `defer c.lock.Unlock()` in `processEntry` and if we did this change we would have to remove that and call before each return.